Octokit.Generators/AsyncPaginationExtensionsGenerator.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System.Linq;
+using System.IO;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using System.Data;
+using System.Text;
+
+namespace Octokit.Generators
+{
+
+  /// <summary>
+  /// AsyncPaginationExtensionsGenerator for generating pagination extensions for Octokit.net Clients that return collections.
+  /// </summary>
+  /// <remarks>
+  /// This generator originally appeared in https://github.com/octokit/octokit.net/pull/2516
+  /// The generator solves a small part of a larger effort that is being discussed:
+  ///  https://github.com/octokit/octokit.net/discussions/2499
+  ///  https://github.com/octokit/octokit.net/discussions/2495
+  ///  https://github.com/octokit/octokit.net/issues/2517
+  ///  In the future, we should be able to unify generation for
+  ///   * models (request and response)
+  ///   * clients
+  ///   * routing and related helpers
+  ///   TODO: Convert to use Rosyln source generators
+  /// </remarks>
+  class AsyncPaginationExtensionsGenerator
+  {
+
+  private const string HEADER = (
+@"using System;
+using System.Collections.Generic;
+
+namespace Octokit.AsyncPaginationExtension
+{
+  /// <summary>
+  /// Provides all extensions for pagination.
+  /// </summary>
+  /// <remarks>
+  /// The <code>pageSize</code> parameter at the end of all methods allows for specifying the amount of elements to be fetched per page.
+  /// Only useful to optimize the amount of API calls made.
+  /// </remarks>
+  public static class Extensions
+  {
+    private const int DEFAULT_PAGE_SIZE = 30;
+  ");
+
+  private const string FOOTER = (
+@"
+  }
+}");
+
+    /// <summary>
+    /// GenerateAsync static entry point for generating pagination extensions.
+    /// </summary>
+    /// <remarks>
+    /// This defaults the search path to the root of the project
+    /// This expects to generate the resulting code and put it in Octokit.AsyncPaginationExtension
+    /// This does a wholesale overwrite on ./Octokit.AsyncPaginationExtension/Extensions.cs
+    /// </remarks>
+    public static async Task GenerateAsync(string root = "./")
+    {
+      var sb = new StringBuilder(HEADER);
+      var enumOptions = new EnumerationOptions { RecurseSubdirectories = true };
+      var paginatedCallRegex = new Regex(@".*Task<IReadOnlyList<(?<returnType>\w+)>>\s*(?<name>\w+)(?<template><.*>)?\((?<arg>.*?)(, )?ApiOptions \w*\);");
+
```

Comment:
What’s the rules on { on same line or new line?

Reply:
We should define that and then add a linter.  I favor concise over verbose unless it makes good sense to be expressive - so if it can be a clean one like expression, why not? That's my opinion, though.

Reply:
I’m happy with anything really as long as it’s the same throughout. I’ve found editorconfig files quite handy but we’d need to make sure it doesn’t fight with Resharper settings 😀